pyro/distributions/lkj.py

```diff
@@ -0,0 +1,197 @@
+        z = x.tanh().clamp(-1 + eps, 1 - eps)
+        return _vector_to_l_cholesky(z)
+
+    def _inverse(self, y):
+        if (y.shape[-2] != y.shape[-1]):
+            raise ValueError("A matrix that isn't square can't be a Cholesky factor of a correlation matrix")
+        D = y.shape[-1]
+
+        z_tri = y.new_zeros(y.shape[:-2] + (D - 2, D - 2))
+        z_stack = [
+            y[..., 1:, 0]
+        ]
+
+        for i in range(2, D):
+            z_tri[..., i - 2, 0:(i-1)] = y[..., i, 1:i] / (1-y[...,i,0:(i-1)].pow(2).cumsum(-1)).sqrt()
+        for j in range(D - 2):
+            z_stack.append(z_tri[..., j:, j])
+
+        z = torch.cat(z_stack, -1)
+        return torch.log1p((2*z)/(1-z)) / 2
+
+    def log_abs_det_jacobian(self, x, y):
+        # Note dependence on pytorch 1.0.1 for batched tril
+        tanpart = x.cosh().log().sum(-1).mul(-2)
+        matpart = (1 - y.pow(2).cumsum(-1).tril(diagonal=-2)).log().div(2).sum(-1).sum(-1)
+        return tanpart + matpart
+
+# register transform to global store
+@biject_to.register(corr_cholesky_constraint)
+@transform_to.register(corr_cholesky_constraint)
+def _transform_to_corr_cholesky(constraint):
+    return CorrLCholeskyTransform()
+
+
+########################################
+# Define distribution
+########################################
+
+# TODO: Modify class to support more than one eta value at a time?
+class CorrLCholeskyLKJPrior(TorchDistribution):
+    """
+    Generates cholesky factors of correlation matrices using an LKJ prior.
+
+    The expected use is to combine it with a vector of variances and pass it
+    to the scale_tril parameter of a multivariate distribution such as MultivariateNormal.
+
+    E.g., if \theta is a (positive) vector of covariances with the same dimensionality
+    as this distribution, and \Omega is sampled from this distribution, scale_tril=torch.mm(torch.diag(sqrt(\theta)), \Omega)
+
+    Note that the `event_shape` of this distribution is `[d, d]`
+
+    :param int d: Dimensionality of the matrix
+    :param torch.Tensor eta: A single positive number parameterizing the distribution.
+    """
+    arg_constraints = {"eta": constraints.positive}
+    support = corr_cholesky_constraint
+    has_rsample = False
+
+    def __init__(self, d, eta):
+        if not torch.is_tensor(eta):
+            raise ValueError("Eta must be a tensor")
+        if any(eta <= 0):
+            raise ValueError("eta must be > 0")
+        vector_size = (d * (d - 1)) // 2
+        alpha = eta.add(0.5 * (d  - 1.0))
+
+        concentrations = eta.new_empty(vector_size,)
+        i = 0
+        for k in range(d-1):
+            alpha -= .5
+            for j in range(k+1, d):
+                concentrations[..., i] = alpha
+                i += 1
+        self._generating_distribution = Beta(concentrations, concentrations)
+        self._eta = eta
+        self._d = d
+        self._lkj_constant = None
+        self._event_shape = torch.Size((d, d))
+
```

Comment:
We will also need to define a `.expand` method which is used to return an expanded distribution instance, and skips some of the parameter checks for efficiency. It used by Pyro to do automatic vectorization. [example code](https://github.com/pytorch/pytorch/blob/master/torch/distributions/lowrank_multivariate_normal.py#L111).

Reply:
Is what's required to support producing multiple samples from the distribution with the same dimensionality and eta?  Or to support producing multiple samples where eta or dimensionality can change?  

If its the latter, that would have a significant performance implication.  The reason is that the bulk of the work in calculating `log_prob` is constant for a given dimensionality and value of `eta`, so we can cache that when its first generated and avoid the repeat work.    

It would also be tricky to support producing samples with different dimensionality in the same batch.

Reply:
Torch distributions have a concept of `batch_shape` and `event_shape` to characterize the shape of samples produced. `event_shape` corresponds to the dependent dimensions of a distribution, i.e. those that are summed out when we call `log_prob`, and the batch dims correspond to the independent dimension (though not necessarily identical) that are retained when we call `log_prob`. 

In this case (like in the multivariate normal case), the event shape is `(d, d)`. This distribution is currently not vectorized, i.e. we can't pass in a tensor of `eta` values, but suppose `eta` could have shape (m, n, 1), where the last dim has size 1 and corresponds to the `eta` for the batch. It would then produce samples of shape (m, n, d, d). Then when we call `.expand([i, j, m, n])`, it will merely expand the `eta` parameters as `eta = eta.expand([i, j, m, n])`. However, this only works if the underlying implementation is batched, like the [multivariate normal](https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py) distribution. I am not sure how feasible that is. Do you think it is possible to batch this distribution and it being able to take in a tensor (of non-identical)`eta`s as mentioned above, and how much work would that be? That will make this distribution play well with Pyro's and PyTorch's batching semantics (and you'll be able to use it inside `pyro.plate` to auto-vectorize).

Reply:
The underlying implementation *is* batched, but only for a single, constant value of `eta`.  

If you parameterize the distribution with `d=3` and `eta=1`, and then call `sample` with a sample size of `i, j, k`, it will return a result of shape `i, j,k,d,d`, where each of the `i * j * k` samples is drawn from the LKJ prior where `eta=1`.   It would not be hard to write an `expand` that increased the batch size but kept `eta` constant.

The difficulty arises when `eta` isn't constant.   

If this is something you feel we need, I will do it.  I would ask you to consider seriously, however --- do we really expect this distribution to be called from inside a `plate`?  Its hard for me to imagine a use-case for that, but I recognize that other people's use-cases are different from my own.

Reply:
Thanks for explaining! I think using this inside `plate` is reasonable, say if you want to specify a prior over the covariance of a batch of multivariate normals. However, it is not obvious to me if there are many cases that need non-identical etas, and I think that's what you were asking. Maybe others can chime in.

To begin with, I think it is reasonable to have a restriction of supporting only iid batching, as you mention. I have the following proposal - that we modify the current constructor to take in tensor `eta` (say with some batch dims `(m, n, ..)`), and we can do a check like `assert eta.max() == eta.min()` in the constructor and throw a `NotImplemented` error for non-identical `eta`. `.expand` would simply involve appending batch dims to `eta` then. If in the future, we see a need to have non-identical eta, the API would remain the same and we can make modifications to the underlying implementation without breaking backward compatibility. Does this sound reasonable?

Reply:
Yes, that makes sense.  So we support varying the `batch_size` while maintaining constant `eta`.  I will implement that tonight and include it in a push with the other issues you've raised.   Thanks!

Reply:
Should be resolved by what I just pushed.